File: fixtures/examples/06-go-stripe/consumer/billing.go
package consumer

import (
	"context"
	"fmt"
)

// Minimal Stripe-shaped client surface for the Go quality harness.
// Mirrors stripe-go CustomerListParams.StartingAfter usage patterns.

type CustomerListParams struct {
	Limit         int64
	StartingAfter string // deprecated — migrate to Page
	Page          string
}

type Customer struct {
	ID    string
	Email string
}

type Client struct {
	Key string
}

// ListCustomers calls the Stripe customers.list equivalent.
func (c *Client) ListCustomers(ctx context.Context, params *CustomerListParams) ([]Customer, error) {
	if params == nil {
		params = &CustomerListParams{Limit: 10}
	}
	// Simulate API call using starting_after query param (legacy).
	_ = params.StartingAfter
	_ = params.Page
	_ = ctx
	if c.Key == "" {
		return nil, fmt.Errorf("missing stripe key")
	}
	return []Customer{{ID: "cus_1", Email: "a@example.com"}}, nil
}

// IterateAllCustomers walks pages using StartingAfter (impact target).
func IterateAllCustomers(ctx context.Context, client *Client) ([]Customer, error) {
	var all []Customer
	var cursor string
	for {
		params := &CustomerListParams{
			Limit:         100,
			StartingAfter: cursor,
		}
		batch, err := client.ListCustomers(ctx, params)
		if err != nil {
			return nil, err
		}
		if len(batch) == 0 {
			break
		}
		all = append(all, batch...)
		cursor = batch[len(batch)-1].ID
		if len(batch) < 100 {
			break
		}
	}
	return all, nil
}

File: fixtures/examples/08-ruby-stripe/consumer/billing.rb
# frozen_string_literal: true

# Minimal Stripe Ruby-shaped client for the quality harness.
# Mirrors Stripe::Customer.list(starting_after: ...) patterns.

module Stripe
  class Customer
    attr_reader :id, :email

    def initialize(id:, email:)
      @id = id
      @email = email
    end

    def self.list(params = {})
      # starting_after is the legacy pagination cursor
      _cursor = params[:starting_after] || params["starting_after"]
      _page = params[:page] || params["page"]
      [Customer.new(id: "cus_1", email: "a@example.com")]
    end
  end
end

module Shop
  class Billing
    def iterate_all_customers
      all = []
      cursor = nil
      loop do
        params = { limit: 100 }
        params[:starting_after] = cursor if cursor
        batch = Stripe::Customer.list(params)
        break if batch.empty?

        all.concat(batch)
        cursor = batch.last.id
        break if batch.length < 100
      end
      all
    end
  end
end
